### PR TITLE
Added new tests for rccl_wrap - rcclUpdateThreadThreshold

### PR DESCRIPTION
**Pull request by awelling2801**
**Work item:** _"Internal"_

**What were the changes?**  
Tests for rcclUpdateThreadThreshold, rcclSetP2pNetChunkSize, and rcclSetPxn as an extension to tests for rccl_wrap.cc - https://github.com/ROCm/rccl/pull/1694

**Why were the changes made?**  
The tests were added as a comment from https://github.com/ROCm/rccl/pull/1694 for a newer function rcclUpdateThreadThreshold

**How was the outcome achieved?**  
7 tests were added to cover all the use cases for rcclUpdateThreadThreshold

**Additional Documentation:**  
The tests should be run as follows - 
```
time ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.*"
time NCCL_THREAD_THRESHOLDS=1 ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.RcclUpdateThreadThreshold_UserEnvSet"
time NCCL_MIN_NCHANNELS=1 ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.RcclUpdateThreadThreshold_MinNChannelsSet"
time NCCL_MAX_NCHANNELS=1 ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.RcclUpdateThreadThreshold_MNChannelsSet"
time NCCL_THREAD_THRESHOLDS=1 NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.RcclUpdateThreadThreshold_UserEnvSet"
time NCCL_MIN_NCHANNELS=1 NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.RcclUpdateThreadThreshold_MinNChannelsSet"
time NCCL_MAX_NCHANNELS=1 NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.RcclUpdateThreadThreshold_MNChannelsSet"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX942_LargeRanks"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX942_BoundaryRank64"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX942_BoundaryRank63"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX950_SmallRanks"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX950_MediumRanks"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX950_LargeRanks"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX950_BoundaryRank16"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX950_BoundaryRank15"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX950_BoundaryRank32"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.GFX950_BoundaryRank31"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.UnsupportedArch_GFX908"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.UnsupportedArch_GFX90A"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.EmptyArchString"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PartialArchMatch"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.ZeroRanks_GFX942"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.ZeroRanks_GFX950"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.LargeRankValues_GFX950"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.CaseInsensitiveArch"
time NCCL_P2P_NET_CHUNKSIZE=123456 NCCL_MAX_NCHANNELS=1 NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.WithEnvironmentVariable"
# Add PXN tests after the existing P2P tests
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_GFX942_SmallRanks"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_GFX942_LargeRanks"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_GFX942_BoundaryRank64"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_GFX942_BoundaryRank63"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_GFX950_SmallRanks"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_GFX950_LargeRanks"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_GFX950_BoundaryRank32"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_GFX950_BoundaryRank31"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_UnsupportedArch_GFX908"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_UnsupportedArch_GFX90A"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_ZeroRanks_GFX942"
time NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_ZeroRanks_GFX950"
time NCCL_PXN_DISABLE=1 NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=ALL ./rccl-UnitTestsFixtures --gtest_filter="Rcclwrap.PXN_WithEnvironmentVariable"
```

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

**Review by rahulvaidya20 — APPROVED** [review]
This looks okay for now. We should also check for edge cases like passing a null ptr, invalid inputs, etc, to some of these functions to ensure they don't break. We can have a different PR for this.

**Comment by venksubramd** [comment]
The tests look good. Do we need tests for other functions in the rccl_wrap.cc file, like rcclSetP2pNetChunkSize for example?

**Comment by awelling2801** [comment]
@venksubramd It is a newer function and can have its own PR later

**Comment by venksubramd** [comment]
> @venksubramd It is a newer function and can have its own PR later

@awelling2801 should we be writing tests for any new function as we write the new functions instead of creating more technical debts along the way?

**Comment by awelling2801** [comment]
> > @venksubramd It is a newer function and can have its own PR later
> 
> @awelling2801 should we be writing tests for any new function as we write the new functions instead of creating more technical debts along the way?

@venksubramd, you are correct. Going forward, the RCCL team should adopt the practice that whenever a new function is added, an accompanying unit test must be written alongside it. This ensures better reliability, easier debugging, and long-term maintainability of the codebase and lesser technical debts along the way.

**Comment by atulkulk** [comment]
> > > @venksubramd It is a newer function and can have its own PR later
> > 
> > 
> > @awelling2801 should we be writing tests for any new function as we write the new functions instead of creating more technical debts along the way?
> 
> @venksubramd, you are correct. Going forward, the RCCL team should adopt the practice that whenever a new function is added, an accompanying unit test must be written alongside it. This ensures better reliability, easier debugging, and long-term maintainability of the codebase and lesser technical debts along the way.

I agree with Anuya. This will then be a never-ending PR. I have added tests for new functions rcclSetP2pNetChunkSize and rcclSetPxn for now. Other new tests for new function(s) should be added as they are added to the main source code as a separate PR.

**Comment by venksubramd** [comment]
> > > > @venksubramd It is a newer function and can have its own PR later
> > > 
> > > 
> > > @awelling2801 should we be writing tests for any new function as we write the new functions instead of creating more technical debts along the way?
> > 
> > 
> > @venksubramd, you are correct. Going forward, the RCCL team should adopt the practice that whenever a new function is added, an accompanying unit test must be written alongside it. This ensures better reliability, easier debugging, and long-term maintainability of the codebase and lesser technical debts along the way.
> 
> I agree with Anuya. This will then be a never-ending PR. I have added tests for new functions rcclSetP2pNetChunkSize and rcclSetPxn for now. Other new tests for new function(s) should be added as they are added to the main source code as a separate PR.

@awelling2801 @atulkulk Moving forward it would be better if we make "having the tests and being reviewed" as part of the definition of done before a ticket is closed.

File: test/RcclWrapTests.cpp [hunk 1]
```diff
@@ -4,124 +4,285 @@
...
-    delete comm;
-}
-
-TEST(Rcclwrap, RcclUpdateCollectiveProtocol_HonorsUserProtocolEnv) {   //Why does this pass if it does not enter the else if block
-  setenv("NCCL_PROTO", "1", 1);  // Simulate manual override
+
+namespace RcclUnitTesting {
+
+// Static flag to ensure only one rcclSetP2pNetChunkSize test runs per execution
+static bool s_p2pNetChunkSizeTestExecuted = false;
```
Comment (PJAvinash) [inline test/RcclWrapTests.cpp:18]: Should this be thread safe ? , If we want to parallelize UT execution going forward.

File: test/RcclWrapTests.cpp [hunk 2]
```diff
@@ -4,124 +4,285 @@
...
+
+// Helper function to check if P2P test should be skipped due to execution order
+static bool ShouldSkipP2pTestDueToExecutionOrder(const std::string &testName) {
```
Comment (PJAvinash) [inline test/RcclWrapTests.cpp:21]: Calling this function twice will have different code paths, In essence this is not idempotent.  Why should we mark s_p2pNetChunkSizeTestExecuted  = true ? should this be Just a read-only function with the debug message? ,  This is against single responsibility pattern.

File: test/RcclWrapTests.cpp [hunk 3]
```diff
@@ -4,124 +4,285 @@
...
+         "  --gtest_filter=Rcclwrap.%s\n"
+         "\nOr run each rcclSetPxn test in separate executions to ensure\n"
+         "proper static variable initialization.\n"
+         "========================\n",
+         testName.c_str(), testName.c_str());
+    return true;
+  }
+
+  // Mark that a PXN test is now executing
+  s_pxnTestExecuted = true;
```
Comment (PJAvinash) [inline test/RcclWrapTests.cpp:96]: function doesn't have Idempotency on same inputs

File: test/RcclWrapTests.cpp [hunk 4]
```diff
@@ -170,4 +332,1359 @@ TEST(Rcclwrap, RcclUpdateCollectiveProtocol_SimpleFallbackWhenNoRanges) {
...
+    INFO(NCCL_LOG_INFO, "[Rcclwrap] Test skipped. Set environment variable "
+                        "NCCL_THREAD_THRESHOLD");
+    GTEST_SKIP() << "[Rcclwrap] Test skipped. Set environment variable "
+                    "NCCL_THREAD_THRESHOLD\n";
+  } else {
+    ncclComm comm = {.nRanks = 8, .nNodes = 4};
+    ncclTaskColl info = {.func = ncclFuncReduceScatter, .protocol = 0};
+    memset(comm.minMaxLLRange, 0, sizeof(comm.minMaxLLRange));
+
+    int threadThreshold = 5; // Any number should do, we should make sure this
```
Comment (PJAvinash) [inline test/RcclWrapTests.cpp:349]: define with const ? or "#define THREAD_THRESHOLD 5" and pass ?

**Comment by PJAvinash** [comment]
int threadThreshold need to have wider variety of inputs  +ve, -ve, zero. 

This PR looks good to me with need for minor improvements.

**Review by PJAvinash — APPROVED** [review]

File: test/RcclWrapTests.cpp (hunk 2 above)
Comment (corey-derochie-amd) [reply test/RcclWrapTests.cpp:21]: I believe this is here due to a weird statefulness in the library preventing certain tests from being run within the same process.

@awelling2801 @atulkulk Any thoughts on somehow running the test in its own process to avoid the statefulness issue? This is a painful restriction if we want to run a whole test suite, and tests are meant to be independent.

File: test/RcclWrapTests.cpp (hunk 1 above)
Comment (atulkulk) [reply test/RcclWrapTests.cpp:18]: It is not thread safe at the moment. We could make it thread safe when we would like to run the tests in parallel. I will create a separate ticket to take a note of this along with other minor issues.

File: test/RcclWrapTests.cpp (hunk 2 above)
Comment (atulkulk) [reply test/RcclWrapTests.cpp:21]: @PJAvinash Yes, calling this function twice will have two different paths and that is the intent of the function in a single execution of these tests. The default execution of the binary will execute only one test. I have included in the INFO message on how to execute these tests. All these tests need to be executed separately (Commands included in PR description).

Comment (atulkulk) [reply test/RcclWrapTests.cpp:21]: @corey-derochie-amd I will look into this.

File: test/RcclWrapTests.cpp (hunk 3 above)
Comment (atulkulk) [reply test/RcclWrapTests.cpp:96]: Correct, hence the tests are required to be run separately as mentioned in the PR description but as Corey highlighted, I will check on a fix.

File: test/RcclWrapTests.cpp (hunk 2 above)
Comment (atulkulk) [reply test/RcclWrapTests.cpp:21]: @corey-derochie-amd @PJAvinash I will have to differ this ask for a separate PR to handle it.

Comment (atulkulk) [reply test/RcclWrapTests.cpp:21]: Created ticket GPUAI-6036.
